File: 01-matrix/main.go
package main

var (
	Dirs = [4][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
)

func maxAreaOfIsland(grid [][]int) int {
	area := 0
	rows := len(grid)
	if rows == 0 {
		return 0
	}
	cols := len(grid[0])
	visited := make([][]bool, rows)
	for v := range visited {
		visited[v] = make([]bool, cols)
	}
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if (grid[i][j] == 1) && !visited[i][j] {
				if a := getAreaDFS(grid, visited, i, j, rows, cols); a > area {
					area = a
				}
			}
		}
	}
	return area
}

func getArea(grid [][]int, visited [][]bool, i, j, rows, cols int) int {
	var pos [2]int
	stack := [][2]int{{i, j}}
	visited[i][j] = true
	area := 0
	for {
		if len(stack) == 0 {
			break
		}
		area++
		pos, stack = stack[0], stack[1:]
		for _, d := range Dirs {
			r := pos[0] + d[0]
			c := pos[1] + d[1]
			if 0 <= r && r < rows && 0 <= c &&
				c < cols && grid[r][c] == 1 &&
				!visited[r][c] {
				stack = append(stack, [2]int{r, c})
				visited[r][c] = true
			}
		}
	}
	return area
}

func getAreaDFS(grid [][]int, visited [][]bool, i, j, rows, cols int) int {
	var area int
	if 0 <= i && i < rows && 0 <= j &&
		j < cols && grid[i][j] == 1 &&
		!visited[i][j] {
		area = 1
	} else {
		return 0
	}
	visited[i][j] = true
	for _, d := range Dirs {
		area += getAreaDFS(grid, visited, i+d[0], j+d[1], rows, cols)
	}
	return area
}

File: 01-matrix/main.py
#!/usr/bin/env python
import unittest
from typing import List, Tuple


"""
https://leetcode.com/problems/01-matrix/
给定一个由 0 和 1 组成的矩阵 mat ，请输出一个大小相同的矩阵，其中每一个格子是 mat 中对应位置元素到最近的 0 的距离。

两个相邻元素间的距离为 1 。

示例 1：

输入：mat = [[0,0,0],[0,1,0],[0,0,0]]
输出：[[0,0,0],[0,1,0],[0,0,0]]
示例 2：

输入：mat = [[0,0,0],[0,1,0],[1,1,1]]
输出：[[0,0,0],[0,1,0],[1,2,1]]
"""
Dirs = ((-1, 0), (1, 0), (0, -1), (0, 1))

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        rows = len(mat)
        if not rows:
            return []
        cols = len(mat[0])
        dists = [[0] * cols for _ in range(rows)]
        stack: List[Tuple[int, int]] = []
        for i in range(rows):
            for j in range(cols):
                if mat[i][j] == 0:
                    dists[i][j] = 0
                    stack.append((i, j))
                else:
                    dists[i][j] = -1
        while stack:
            i, j= stack.pop()
            dist = dists[i][j]
            for d in Dirs:
                r = i+
        return []


cases = [
    {
        "input": [[[0, 0, 0], [0, 1, 0], [0, 0, 0]]],
        "output": [[0, 0, 0], [0, 1, 0], [0, 0, 0]],
    },
    {
        "input": [[[0, 0, 0], [0, 1, 0], [1, 1, 1]]],
        "output": [[0, 0, 0], [0, 1, 0], [1, 2, 1]],
    },
]


class SolutionTestCase(unittest.TestCase):
    def test(self):
        for t in cases:
            print(f"input: {t['input']}\noutput: {t['output']}\n")
            ret = Solution().updateMatrix(*t["input"])
            self.assertListEqual(ret, t["output"])


if __name__ == "__main__":
    unittest.main()
